### ISSUE TITLE
`meteor test-packages` fails when `Npm.depends` is nested _only_ in `onUse`

### ISSUE THREAD

**Issue by harishreddy-m** [meteor/meteor#8200]
Meteor version : 1.4.2.3
OS : elementary OS (based on ubuntu 16.04)

I am trying to write a meteor package.I want to use a npm module as dependency for the package.I am following the [guide](https://guide.meteor.com/writing-atmosphere-packages.html#npm-dependencies)

When I run `meteor test-packages --driver-package practicalmeteor:mocha ./`
I am getting error stack, appended at the end.
Tests are running fine if I remove Npm.depends()

one solution/hack I tried -
 create a package.json,
write npm module dependencies in package.json 
run meteor npm install --> It created node_modules folder at root level and imports in code are working.

But I prefer Npm.depends({}) as it is clean approach.

```
updating npm dependencies -- github...
/home/harishreddy/.meteor/packages/ecmascript/.0.5.9.ui3230++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:190
      throw error;
      ^

Error: ENOENT: no such file or directory, lstat '/home/harishreddy/meteorPackages/communication/.npm/package/node_modules'
    at Error (native)
    at Object.fs.lstatSync (fs.js:839:18)
    at Object.wrapper (/tools/fs/files.js:1535:35)
    at files.(anonymous function) (/tools/fs/files.js:1636:27)
    at wrap.makeCacheKey._len2 (/tools/fs/optimistic.js:22:15)
    at Entry._reallyRecompute (/home/harishreddy/.meteor/packages/meteor-tool/.1.4.2_3.h5rt9d++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/optimism/lib/entry.js:215:26)
    at Entry._recomputeIfDirty (/home/harishreddy/.meteor/packages/meteor-tool/.1.4.2_3.h5rt9d++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/optimism/lib/entry.js:179:17)
    at Entry.recompute (/home/harishreddy/.meteor/packages/meteor-tool/.1.4.2_3.h5rt9d++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/optimism/lib/entry.js:46:20)
    at optimistic (/home/harishreddy/.meteor/packages/meteor-tool/.1.4.2_3.h5rt9d++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/optimism/lib/index.js:50:18)
    at /tools/isobuild/meteor-npm.js:438:17
    at Object.meteorNpm.dependenciesArePortable (/tools/isobuild/meteor-npm.js:520:10)
    at NodeModulesDirectory.isPortable (/tools/isobuild/bundler.js:269:22)
    at /tools/isobuild/bundler.js:1975:17
    at Function._.each._.forEach (/home/harishreddy/.meteor/packages/meteor-tool/.1.4.2_3.h5rt9d++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:1972:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2241:13)
    at /tools/isobuild/bundler.js:2336:30
    at /tools/isobuild/bundler.js:2469:9
    at Array.forEach (native)
    at /tools/isobuild/bundler.js:2459:26
    at /tools/isobuild/bundler.js:2399:7
    at /tools/isobuild/bundler.js:2800:22
    at /tools/utils/buildmessage.js:271:13
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2633:31)
    at /tools/isobuild/bundler.js:2580:32
    at Object.withCache (/tools/fs/files.js:1601:12)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2580:16)
    at /tools/runners/run-app.js:574:36
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:573:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:617:35)
    at AppRunner._fiber (/tools/runners/run-app.js:876:28)
    at /tools/runners/run-app.js:403:12
```

**Comment by abernix** [comment]
Please attach a reproduction repo per the guidelines for [reporting a bug](https://github.com/meteor/meteor/blob/devel/Contributing.md#reporting-bug).

**Comment by abernix** [comment]
Thanks for providing the reproduction!  I can confirm that this is happening in your reproduction, but I think it's important to point out a significant detail between your two reproductions, and also the way it's explained in the guide.  The [Guide states](https://guide.meteor.com/writing-atmosphere-packages.html#npm-dependencies):

> Use [`Npm.depends`](http://docs.meteor.com/#/full/Npm-depends) at the top level of your `package.js` file.

In your reproduction, you have it _nested_ inside of `Package.onUse()` – an important distinction!  In your "OR" reproduction steps, you don't specify adding `Npm.depends` inside `Package.onUse()`.  If you simply add it to the top-level of the `package.js`, you will find that your reproduction works just fine.

You can skip the rest of this update, unless you're really interested.  I'll leave this open for now.

---

I definitely don't believe it should error in this way, however allowing `onTest` and `onUse` to  have differentiating `node_modules` is currently not supported, [as indicated by this comment](https://github.com/meteor/meteor/blob/c29a1e9b0be6f7cb0472eaa96213b99b5530bff0/tools/isobuild/package-npm.js#L50-L52).

For a complicated explanation (mainly this is just going to be notes from my own exercise in looking at this, but hopefully helpful for someone 😬), it should first be pointed out that the `Package.onUse` and `Package.onTest` components are built separately – but one uses the other.  To work properly, the `Package.onTest` `use`s the package (isopack) from the `Package.onUse` portion.

In the case of this reproduction, the initial compilation of the `Package.onUse` portion (which is done first) is generating the `.npm/package/node_modules` directory based on the `Npm.depends` which you have in the `onUse` clause (so, `github@3.1.1`) inside the `meteor-package-with-npm-dep` package.

When it comes time to run the `Package.onTest` block, Meteor utilizes a special bundling of what I might call a pseudo-package – essentially, it makes a package by the same name, but prefixed with `local-test:` (This is done invisibly to the developer).   Meteor shares some of the package build environment when building the pseudo-package, but it does take two passes, one normal, and one pseudo.

On the first pass (for `onRun`), `Npm.depends` is set however on the second pass (for `onTest`), since `Npm.depends` isn't inside the `onTest`, while it's [updating dependencies](https://github.com/meteor/meteor/blob/c29a1e9b0be6f7cb0472eaa96213b99b5530bff0/tools/isobuild/compiler.js#L126-L128) the packages are [removed](https://github.com/meteor/meteor/blob/c29a1e9b0be6f7cb0472eaa96213b99b5530bff0/tools/isobuild/meteor-npm.js#L70-L87).

This would be mostly okay, however Meteor's bundler doesn't quite take proper inventory of that and includes it in the bundle later on.  If it were to check [somewhere in here, perhaps](https://github.com/meteor/meteor/blob/c29a1e9b0be6f7cb0472eaa96213b99b5530bff0/tools/isobuild/bundler.js#L1882-L1884) if the directories existed, the error could be avoided – somewhat...

If the `Npm.depends` was only defined in the `onUse`, then this pattern _might_ work, though I haven't evaluated extensively.  Presumably, you wouldn't have access to the modules from your test code – which would be expected.

I believe a problem would certainly arise right now if you try using differing `Npm.depends` in both `onUse` and `onTest` though.  I _think_ this would cause a loop of the builder alternating between building the test packages and the regular packages.  Going back to the [code I referenced earlier]([as indicated by this comment](https://github.com/meteor/meteor/blob/c29a1e9b0be6f7cb0472eaa96213b99b5530bff0/tools/isobuild/package-npm.js#L50-L52)), I would actually have thought this would have been prevented, however that doesn't seem to be stopping this pattern right now.  This would need some further research to really know what's going on.

**tl;dr**: Remember that `Npm.depends` is less ideal in more recent versions of Meteor thanks to the ability to use [peer NPM dependencies](https://guide.meteor.com/writing-atmosphere-packages.html#peer-npm-dependencies).

See also: #6888

**Comment by hwillson** [comment]
Hi all - as mentioned previously in this issue thread, this problem is caused by using `Npm.depends` within the `onTest` function (for more specific details, see @abernix's excellent explanation in https://github.com/meteor/meteor/issues/8200#issuecomment-270723863). The solution to this problem is to use `Npm.depends` outside of `onTest` / `onUse`, at the top level of the `package.js` file. This isn't really a workaround - it's actually the recommended approach when using `Npm.depends` (and is documented in the [npm dependencies](https://guide.meteor.com/writing-atmosphere-packages.html#npm-dependencies) section of the Guide). Yes the Package API could be updated to prevent the use of `Npm.depends` within `onTest` / `onUse` (by throwing an error for example), but given that using `Npm.depends` is no longer the recommended approach for handling peer npm dependencies with Meteor (as documented in the [Peer npm dependencies](https://guide.meteor.com/writing-atmosphere-packages.html#peer-npm-dependencies) section of the Guide), the chances of this issue being worked on further are pretty low. Give this, and the fact that anyone encountering this issue will be able to find this issue thread when searching the GH issue archive, I'll close this issue. Happy to re-open if anyone disagrees. Thanks for reporting this issue originally!
